x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -354,34 +500,214 @@ public boolean equals(Object o) {
             RealmRef realmRef = (RealmRef) o;
 
             if (nodeName.equals(realmRef.nodeName) == false) return false;
-            if (name.equals(realmRef.name) == false) return false;
-            return type.equals(realmRef.type);
+            if (type.equals(realmRef.type) == false) return false;
+            return Objects.equals(domain, realmRef.domain);
         }
 
         @Override
         public int hashCode() {
             int result = nodeName.hashCode();
             result = 31 * result + name.hashCode();
             result = 31 * result + type.hashCode();
+            if (domain != null) {
+                result = 31 * result + domain.hashCode();
+            }
             return result;
         }
 
         @Override
         public String toString() {
-            return "{Realm[" + type + "." + name + "] on Node[" + nodeName + "]}";
+            if (domain != null) {
+                return "{Realm[" + type + "." + name + "] under Domain[" + domain.name() + "] on Node[" + nodeName + "]}";
+            } else {
+                return "{Realm[" + type + "." + name + "] on Node[" + nodeName + "]}";
+            }
+        }
+
+        static RealmRef newInternalAttachRealmRef(String nodeName) {
+            // the "attach" internal realm is not part of any realm domain
+            return new Authentication.RealmRef(ATTACH_REALM_NAME, ATTACH_REALM_TYPE, nodeName, null);
+        }
+
+        static RealmRef newInternalFallbackRealmRef(String nodeName) {
+            // the "fallback" internal realm is not part of any realm domain
+            RealmRef realmRef = new RealmRef(FALLBACK_REALM_NAME, FALLBACK_REALM_TYPE, nodeName, null);
+            return realmRef;
+        }
+
+        static RealmRef newAnonymousRealmRef(String nodeName) {
+            // the "anonymous" internal realm is not part of any realm domain
+            return new Authentication.RealmRef(ANONYMOUS_REALM_NAME, ANONYMOUS_REALM_TYPE, nodeName, null);
+        }
+
+        static RealmRef newServiceAccountRealmRef(String nodeName) {
+            // no domain for service account tokens
+            return new Authentication.RealmRef(ServiceAccountSettings.REALM_NAME, ServiceAccountSettings.REALM_TYPE, nodeName, null);
+        }
+
+        static RealmRef newApiKeyRealmRef(String nodeName) {
+            // no domain for API Key tokens
+            return new RealmRef(AuthenticationField.API_KEY_REALM_NAME, AuthenticationField.API_KEY_REALM_TYPE, nodeName, null);
         }
     }
 
-    public static ConstructingObjectParser<RealmRef, Void> REALM_REF_PARSER = new ConstructingObjectParser<>(
-        "realm_ref",
-        false,
-        (args, v) -> new RealmRef((String) args[0], (String) args[1], (String) args[2])
-    );
+    // TODO is a newer version than the node's a valid value?
+    public static Authentication newInternalAuthentication(User internalUser, Version version, String nodeName) {
```

Comment:
I'm not a fan of the "attach" designation used for the realm of the internal users, I don't want it to propagate because I don't see it as relevant. "Internal" authentication from "internal" users is clear enough, IMO.